xds/internal/balancer/ringhash/picker.go

```diff
@@ -34,32 +37,87 @@ type picker struct {
+		var ok bool
+		if requestHash, ok = GetXDSRequestHash(info.Ctx); !ok {
+			return balancer.PickResult{}, fmt.Errorf("bug: expected xDS config selector to set the request hash")
+		}
+	} else {
+		md, ok := metadata.FromOutgoingContext(info.Ctx)
+		if !ok || len(md.Get(p.requestHashHeader)) == 0 {
+			requestHash = p.randUint64()
+			usingRandomHash = true
+		} else {
+			header := strings.Join(md.Get(p.requestHashHeader), ",")
+			requestHash = xxhash.Sum64String(header)
+		}
+	}
+
+	e := p.ring.pick(requestHash)
 	ringSize := len(p.ring.items)
-	// Per gRFC A61, because of sticky-TF with PickFirst's auto reconnect on TF,
-	// we ignore all TF subchannels and find the first ring entry in READY,
-	// CONNECTING or IDLE.  If that entry is in IDLE, we need to initiate a
-	// connection. The idlePicker returned by the LazyLB or the new Pickfirst
-	// should do this automatically.
-	for i := 0; i < ringSize; i++ {
-		index := (e.idx + i) % ringSize
-		balState := p.balancerState(p.ring.items[index])
-		switch balState.ConnectivityState {
-		case connectivity.Ready, connectivity.Connecting, connectivity.Idle:
-			return balState.Picker.Pick(info)
-		case connectivity.TransientFailure:
-		default:
-			panic(fmt.Sprintf("Found child balancer in unknown state: %v", balState.ConnectivityState))
+	if !usingRandomHash {
+		// Per gRFC A61, because of sticky-TF with PickFirst's auto reconnect on TF,
+		// we ignore all TF subchannels and find the first ring entry in READY,
+		// CONNECTING or IDLE.  If that entry is in IDLE, we need to initiate a
+		// connection. The idlePicker returned by the LazyLB or the new Pickfirst
+		// should do this automatically.
+		for i := 0; i < ringSize; i++ {
+			index := (e.idx + i) % ringSize
+			balState := p.balancerState(p.ring.items[index])
+			switch balState.ConnectivityState {
+			case connectivity.Ready, connectivity.Connecting, connectivity.Idle:
+				return balState.Picker.Pick(info)
+			case connectivity.TransientFailure:
+			default:
+				panic(fmt.Sprintf("Found child balancer in unknown state: %v", balState.ConnectivityState))
+			}
+		}
+	} else {
+		// If the picker has generated a random hash, it will walk the ring from
+		// this hash, and pick the first READY endpoint. If no endpoint is
+		// currently in CONNECTING state, it will trigger a connection attempt
+		// on at most one endpoint that is in IDLE state along the way. - A76
+		requestedConnection := p.hasEndpointInConnectingState
+		for i := 0; i < ringSize; i++ {
+			index := (e.idx + i) % ringSize
+			balState := p.balancerState(p.ring.items[index])
+			if balState.ConnectivityState == connectivity.Ready {
+				return balState.Picker.Pick(info)
+			}
+			if !requestedConnection && balState.ConnectivityState == connectivity.Idle {
+				requestedConnection = true
+				// If the picker is in idle state, initiate a connection but
+				// continue to check other pickers to see if there is one in
+				// ready state.
+				balState.Picker.Pick(info)
```

Comment:
Do we rely on this `Pick` call to initiate the connection? The `endpointStates` map in the `ringhashBalancer` does have the underlying child balancer on which we could call `ExitIdle` to initiate a connection, but that information is not shared with the picker (only the balancer state is shared, which contains the connectivity state and the underlying picker).

Maybe I'm missing something.

Reply:
We discussed this yesterday, you should be able to call `ExitIdle` on the Balancer provided by endpointsharding instead of calling `Pick`.
https://github.com/grpc/grpc-go/blob/5edab9e55414068e74320716117a2659c5d2174e/xds/internal/balancer/ringhash/ringhash.go#L250-L262

We don't want to use the picker unless an RPC is going to be made.